File: mobile/lib/models/api/user/srp.dart
class SetupSRPRequest {
  final String srpUserID;
  final String srpSalt;
  final String srpVerifier;
  final String srpA;
  final bool isUpdate;

  SetupSRPRequest({
    required this.srpUserID,
    required this.srpSalt,
    required this.srpVerifier,
    required this.srpA,
    required this.isUpdate,
  });

  Map<String, dynamic> toMap() {
    return {
      'srpUserID': srpUserID.toString(),
      'srpSalt': srpSalt,
      'srpVerifier': srpVerifier,
      'srpA': srpA,
      'isUpdate': isUpdate,
    };
  }

  factory SetupSRPRequest.fromJson(Map<String, dynamic> json) {
    return SetupSRPRequest(
      srpUserID: json['srpUserID'],
      srpSalt: json['srpSalt'],
      srpVerifier: json['srpVerifier'],
      srpA: json['srpA'],
      isUpdate: json['isUpdate'],
    );
  }
}

class SetupSRPResponse {
  final String setupID;
  final String srpB;

  SetupSRPResponse({
    required this.setupID,
    required this.srpB,
  });

  Map<String, dynamic> toMap() {
    return {
      'setupID': setupID.toString(),
      'srpB': srpB,
    };
  }

  factory SetupSRPResponse.fromJson(Map<String, dynamic> json) {
    return SetupSRPResponse(
      setupID: json['setupID'],
      srpB: json['srpB'],
    );
  }
}

class CompleteSRPSetupRequest {
  final String setupID;
  final String srpM1;

  CompleteSRPSetupRequest({
    required this.setupID,
    required this.srpM1,
  });

  Map<String, dynamic> toMap() {
    return {
      'setupID': setupID.toString(),
      'srpM1': srpM1,
    };
  }

  factory CompleteSRPSetupRequest.fromJson(Map<String, dynamic> json) {
    return CompleteSRPSetupRequest(
      setupID: json['setupID'],
      srpM1: json['srpM1'],
    );
  }
}

class SrpAttributes {
  final String srpUserID;
  final String srpSalt;
  final int memLimit;
  final int opsLimit;
  final String kekSalt;
  final bool isEmailMFAEnabled;

  SrpAttributes({
    required this.srpUserID,
    required this.srpSalt,
    required this.memLimit,
    required this.opsLimit,
    required this.kekSalt,
    required this.isEmailMFAEnabled,
  });

  factory SrpAttributes.fromMap(Map<String, dynamic> map) {
    return SrpAttributes(
      srpUserID: map['attributes']['srpUserID'],
      srpSalt: map['attributes']['srpSalt'],
      memLimit: map['attributes']['memLimit'],
      opsLimit: map['attributes']['opsLimit'],
      kekSalt: map['attributes']['kekSalt'],
      isEmailMFAEnabled: map['attributes']['isEmailMFAEnabled'],
    );
  }
}

class CompleteSRPSetupResponse {
  final String setupID;
  final String srpM2;

  CompleteSRPSetupResponse({
    required this.setupID,
    required this.srpM2,
  });

  Map<String, dynamic> toMap() {
    return {
      'setupID': setupID,
      'srpM2': srpM2,
    };
  }

  factory CompleteSRPSetupResponse.fromJson(Map<String, dynamic> json) {
    return CompleteSRPSetupResponse(
      setupID: json['setupID'],
      srpM2: json['srpM2'],
    );
  }
}

File: server/ente/srp.go
package ente

import (
	"github.com/google/uuid"
)

type SetupSRPRequest struct {
	SrpUserID   uuid.UUID `json:"srpUserID" binding:"required"`
	SRPSalt     string    `json:"srpSalt" binding:"required"`
	SRPVerifier string    `json:"srpVerifier" binding:"required"`
	SRPA        string    `json:"srpA"  binding:"required"`
}

type SetupSRPResponse struct {
	SetupID uuid.UUID `json:"setupID" binding:"required"`
	SRPB    string    `json:"srpB" binding:"required"`
}

type CompleteSRPSetupRequest struct {
	SetupID uuid.UUID `json:"setupID" binding:"required"`
	SRPM1   string    `json:"srpM1" binding:"required"`
}

type CompleteSRPSetupResponse struct {
	SetupID uuid.UUID `json:"setupID" binding:"required"`
	SRPM2   string    `json:"srpM2" binding:"required"`
}

type RecoverySrpSetupRequest struct {
	RecoveryID  uuid.UUID       `json:"recoveryID" binding:"required"`
	SetUpSRPReq SetupSRPRequest `json:"setupSRPRequest" binding:"required"`
}

type RecoveryUpdateSRPAndKeysRequest struct {
	RecoveryID uuid.UUID               `json:"recoveryID" binding:"required"`
	UpdateSrp  UpdateSRPAndKeysRequest `json:"updateSrpAndKeysRequest" binding:"required"`
}

// UpdateSRPAndKeysRequest is used to update the SRP attributes (e.g. when user updates his password) and also
// update the keys attributes
type UpdateSRPAndKeysRequest struct {
	SetupID            uuid.UUID          `json:"setupID" binding:"required"`
	SRPM1              string             `json:"srpM1" binding:"required"`
	UpdateAttributes   *UpdateKeysRequest `json:"updatedKeyAttr"`
	LogOutOtherDevices *bool              `json:"logOutOtherDevices"`
}

type UpdateSRPSetupResponse struct {
	SetupID uuid.UUID `json:"setupID" binding:"required"`
	SRPM2   string    `json:"srpM2" binding:"required"`
}

type GetSRPAttributesRequest struct {
	Email string `form:"email" binding:"required"`
}

type GetSRPAttributesResponse struct {
	SRPUserID string `json:"srpUserID" binding:"required"`
	SRPSalt   string `json:"srpSalt"  binding:"required"`
	// MemLimit,OpsLimit,KekSalt are needed to derive the KeyEncryptionKey
	// on the client. Client generates the LoginKey from the KeyEncryptionKey
	// and treat that as UserInputPassword.
	MemLimit          int    `json:"memLimit" binding:"required"`
	OpsLimit          int    `json:"opsLimit" binding:"required"`
	KekSalt           string `json:"kekSalt" binding:"required"`
	IsEmailMFAEnabled bool   `json:"isEmailMFAEnabled" binding:"required"`
}

type CreateSRPSessionRequest struct {
	SRPUserID uuid.UUID `json:"srpUserID" binding:"required"`
	SRPA      string    `json:"srpA" binding:"required"`
}

type CreateSRPSessionResponse struct {
	SessionID uuid.UUID `json:"sessionID" binding:"required"`
	SRPB      string    `json:"srpB" binding:"required"`
}

type VerifySRPSessionRequest struct {
	SessionID uuid.UUID `json:"sessionID" binding:"required"`
	SRPUserID uuid.UUID `json:"srpUserID" binding:"required"`
	SRPM1     string    `json:"srpM1"`
}

// SRPSessionEntity represents a row in the srp_sessions table
type SRPSessionEntity struct {
	ID           uuid.UUID
	SRPUserID    uuid.UUID
	UserID       int64
	ServerKey    string
	SRP_A        string
	IsVerified   bool
	AttemptCount int32
	IsFake       bool
}

type SRPAuthEntity struct {
	UserID    int64
	SRPUserID uuid.UUID
	Salt      string
	Verifier  string
}

type SRPSetupEntity struct {
	ID        uuid.UUID
	SessionID uuid.UUID
	SRPUserID uuid.UUID
	UserID    int64
	Salt      string
	Verifier  string
}
